<a href="https://colab.research.google.com/github/1st-award/andong_2021_2_1/blob/main/Python/%ED%95%84%EA%B8%B0%20%EC%82%AC%EB%B3%B8/%ED%9A%8C%EA%B7%80%EB%B6%84%EC%84%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 회귀분석(Regression Analysis)

**회귀분석**: 독립변수(x)와 종속변수(y) 사이으이 관계를 찾는 통계적 방법
**선형회귀분석(단순회귀분석)**: 독립변수와 종속변수 사이에 **직선적인 형태의 관계**가 있다고 가정하는 것

## 독립변수와 종속변수
- 독립변수(independent variable)또는 설명변수(explanatory variable): 두 변수의 인과관계에서 영향을 주는 변수, 즉 원인이 되는 변수(X)
- 종속변수(dependent variable)또는 반응변수(response variable): 두 변수 중 다른 변수의 변화에 따라 값이 변화하는 변수(Y)

회귀분석에 대한 자세한 내용은 첨부한 강의자료를 참조하세요.


## 간단한 회귀분석 예

In [ ]:
from google.colab import drive
drive.mount('/gdrive', force_remount=True)

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

info = pd.read_excel('/gdrive/My Drive/Python/2015_korean_data.xlsx')
info.head() # head는 가장 위에 있는 5개 출력 반대로 tail이 있다.

## (1) 키와 몸무게 사이의 관계 분석

In [ ]:
# 남자 데이터 추출
is_male = info['성별'] == '남'
male_info = info[is_male]
# 여자 데이터 추출
is_female = info['성별'] == '여'
female_info = info[is_female]

In [ ]:
# 데이터프레임 정보 확인 - 결측치 포함 여부 확인
male_info.info()
female_info.info()

In [ ]:
# 결측치가 포함된 행 전체 삭제
# 남자 결측치 제거
male_info = male_info.dropna(axis=0)
male_info.info()
# 여자 결측치 제거
female_info = female_info.dropna(axis=0)
female_info.info()

In [ ]:
# 남자의 키와 몸무게 사이의 관계 분석
x = male_info['키']
y = male_info['몸무게']
plt.scatter(x, y, alpha=0.5)
plt.title("Analysis of hegiht and weight relationship(male)")
plt.xlabel("height")
plt.ylabel("weight")
plt.grid()
plt.show()

### 선형회귀식 추정

In [ ]:
# 1. 남자 키와 몸무게 사이의 회귀분석
from sklearn.linear_model import LinearRegression
fitter = LinearRegression()
fitter.fit(x.values.reshape(-1, 1), y)
b1 = fitter.coef_
b0 = fitter.intercept_
print(b1, b0)

In [ ]:
# 2. 남자의 키와 몸무게 사이의 회귀분석 결과의 시각화
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
info = pd.read_excel('/gdrive/My Drive/Python/2015_korean_data.xlsx')

# 남자 데이터 추출
is_male = info['성별'] == '남'
male_info = info[is_male]
male_info = male_info.dropna(axis=0)

x = male_info['키']
y = male_info['몸무게']
fitter = LinearRegression()
fitter.fit(x.values.reshape(-1, 1), y) # 직선 추정
y_h = fitter.predict(x.values.reshape(-1, 1)) # 추정된 직선의 방정식
plt.scatter(x, y, color="slateblue", alpha=0.6) # 산점도 그리기
plt.plot(x, y_h, color='red') # 추정된 직선 그리기
plt.title("Analyis of height and weight relationship(male)")
plt.xlabel("height")
plt.ylabel("weight")
plt.grid() # 격자무늬 표시
plt.show()



[과제1] 여자의 키와 몸무게 사이의 선형회귀분석을 실시하고 그 결과를 시각화하시오

In [ ]:
# [과제1] 여자의 키, 몸무게 사이의 회귀분석
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
info = pd.read_excel('/gdrive/My Drive/Python/2015_korean_data.xlsx')

# 여자 데이터 추출
is_female = info['성별'] == '여'
female_info = info[is_female]
female_info = female_info.dropna(axis=0)

x = female_info['키']
y = female_info['몸무게']
fitter = LinearRegression()
fitter.fit(x.values.reshape(-1, 1), y) # 직선 추정
y_h = fitter.predict(x.values.reshape(-1, 1)) # 추정된 직선의 방정식
plt.scatter(x, y, color="slateblue", alpha=0.6) # 산점도 그리기
plt.plot(x, y_h, color='red') # 추정된 직선 그리기
plt.title("Analyis of height and weight relationship(male)")
plt.xlabel("height")
plt.ylabel("weight")
plt.grid() # 격자무늬 표시
plt.show()

[과제2] 흔히 발이 크면 키도 크다는 속설이 있습니다. 남자 데이터프레임을 이용하여 발 길이(독립변수 x)와 키(종속변수, y)사이의 회귀분석을 시행하고 그 결과를 시각화하시오.

In [ ]:
# [과제2] 남자의 발. 키 사이의 회귀분석
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
info = pd.read_excel('/gdrive/My Drive/Python/2015_korean_data.xlsx')

# 남자 데이터 추출
is_male = info['성별'] == '남'
male_info = info[is_male]
male_info = male_info.dropna(axis=0)

x = male_info['발길이']
y = male_info['키']
fitter = LinearRegression()
fitter.fit(x.values.reshape(-1, 1), y) # 직선 추정
y_h = fitter.predict(x.values.reshape(-1, 1)) # 추정된 직선의 방정식
plt.scatter(x, y, color="slateblue", alpha=0.6) # 산점도 그리기
plt.plot(x, y_h, color='red') # 추정된 직선 그리기
plt.title("Analyis of foot length and height relationship(male)")
plt.xlabel("foot length")
plt.ylabel("height")
plt.grid() # 격자무늬 표시
plt.show()


## 결정계수(R<sup>2</sup>) 의 계산

결정계수(Coefficient of determination)
- 추정한 선형모델(선형회귀직선)이 주어진 자료에 적합한 정도를 나타내는 척도
- 추정한 선형모델의 설명력
- 결정계수의 범위: 0부터 1사이의 실수 값
    - 결정계수의 값이 1에 가까울수록 추정한 선형회귀직선의 유용성이 높고, 0에 가까울수록 유용성이 낮다

### 결정계수(R<sup>2</sup>)를 이용한 남자 키와 몸무게 사잉의 회귀 분석 검증

In [ ]:
# sklearn 라이브러리를 이용한 결정 계수 계산
from sklearn.metrics import r2_score
r2 = r2_score(male_info['몸무게'], y_h)
r2

In [ ]:
# yi의 추정치 계산
f = np.polyfit(x, y, 1) # x = male_info['키'], y = male_info['몸무게']
f = np.poly1d(f)
x = np.array(x) # y = f(x)
# SST = np.sum((male_info['몸무게']- exp_B)**2)
SST = np.sum((y-np.mean(y))**2)
SSR = np.sum((f(x)-np.mean(y))**2)
R2 = SSR/SST
R2

0.16137317613311758

결론: 결정계수의 값이 약 0.16137으로 추정된 회귀직선이 유의밍하게 키와 몸무게 사이의 관계를 표현한다고 보기 어렵다.

### 머리 크기와 몸무게 사이의 회귀분석 및 검증 - 여자 데이터 이용

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# 파일 읽기
info = pd.read_excel('/gdrive/My Drive/Python/2015_korean_data.xlsx')

# 여자 데이터 추출
is_female = info['성별'] == '여'
female_info = info[is_female]
female_info = female_info.dropna(axis=0) # 결측치가 포함된 행 전체 삭제

x = female_info['머리둘레']
y = female_info['몸무게']
fitter = LinearRegression()
fitter.fit(x.values.reshape(-1, 1), y)
y_h = fitter.predict(x.values.reshape(-1, 1)) # 추정된 선형회귀 직선
plt.scatter(x, y, color='slateblue', alpha=0.6)
plt.plot(x, y_h, color='red')
plt.title("Analysis of head size and weight relationship(female)")
plt.xlabel("head size")
plt.ylabel("weight")
plt.grid()
plt.show()

In [ ]:
# 결정계수를 이용한 회귀분석 결과 검증
r2 = r2_score(female_info['몸무게'], y_h)
r2

In [ ]:
# yi의 추정치 계산
f = np.polyfit(female_info['머리둘레'], female_info['몸무게'], 1)
f = np.poly1d(f)

SST = np.sum((y - np.mean(y))**2)
SSR = np.sum((f(x) - np.mean(y))**2)
R2 = SSR / SST
R2

0.11981438698325955

결론: (여자) 머리둘레의 길이와 몸무게 사이의 관계를 추정하는 선형회귀분석 모델이 데이터를 잘 설명하기 어렵다.